From b28ccf6f057e1412a8e02ebb3e12deb5c1bd74bb Mon Sep 17 00:00:00 2001
From: David Ferguson <ferguson.david@gmail.com>
Date: Thu, 7 Jun 2012 21:00:41 -0400
Subject: [PATCH] ext4_utils: add BoardConfig define to suppress
 EMMC-corrupting wipe command

If BOARD_SUPPRESS_EMMC_WIPE is true, the EMMC wipe command will not be
issued. This works around a bug in some firmware revisions of Samsung EMMC's
that permanently damages the device when the wipe command is issued.

For affected devices with kernel source, MMC_CAP_ERASE should be removed
from the kernel instead.

This is only part of the solution but it does handle the "flashing CM9
for the first time on an unsafe kernel" situation.

Change-Id: Ie4e31f9268a65218e5d344ae3068b021790fc33c
---
 ext4_utils/Android.mk | 10 ++++++++++
 ext4_utils/wipe.c     |  8 ++++++++
 2 files changed, 18 insertions(+)

diff --git a/ext4_utils/Android.mk b/ext4_utils/Android.mk
index c5684f9..b69b2ac 100644
--- a/ext4_utils/Android.mk
+++ b/ext4_utils/Android.mk
@@ -61,6 +61,11 @@ LOCAL_SHARED_LIBRARIES := \
     libselinux \
     libsparse \
     libz
+
+ifeq ($(BOARD_SUPPRESS_EMMC_WIPE),true)
+    LOCAL_CFLAGS += -DSUPPRESS_EMMC_WIPE
+endif
+
 include $(BUILD_SHARED_LIBRARY)
 
 
@@ -70,6 +75,11 @@ LOCAL_MODULE := libext4_utils_static
 LOCAL_STATIC_LIBRARIES += \
     libselinux \
     libsparse_static
+
+ifeq ($(BOARD_SUPPRESS_EMMC_WIPE),true)
+    LOCAL_CFLAGS += -DSUPPRESS_EMMC_WIPE
+endif
+
 include $(BUILD_STATIC_LIBRARY)
 
 
diff --git a/ext4_utils/wipe.c b/ext4_utils/wipe.c
index 5766632..4b6933d 100644
--- a/ext4_utils/wipe.c
+++ b/ext4_utils/wipe.c
@@ -32,6 +32,7 @@
 #define BLKSECDISCARD _IO(0x12,125)
 #endif
 
+#ifndef SUPPRESS_EMMC_WIPE
 int wipe_block_device(int fd, s64 len)
 {
 	u64 range[2];
@@ -60,6 +61,13 @@ int wipe_block_device(int fd, s64 len)
 
 	return 0;
 }
+#else
+int wipe_block_device(int fd, s64 len)
+{
+	warn("Wipe via secure discard suppressed due to bug in EMMC firmware\n");
+	return 1;
+}
+#endif /* SUPPRESS_EMMC_WIPE */
 
 #else  /* __linux__ */
 #error "Missing block device wiping implementation for this platform!"
-- 
2.7.4

